packages/internal/src/lib-nodejs/ava-unhandled-rejection.js

```diff
@@ -14,6 +14,104 @@ export const AVA_EXPECT_UNHANDLED_REJECTIONS =
 
 export const SUBTEST_PREFIX = '(unhandled rejection subprocess): ';
 
+const delayTurn = () => new Promise(resolve => setImmediate(resolve));
+const settleUnhandled = async () => {
+  await delayTurn();
+  await delayTurn();
+};
+
+const stripListeners = event => {
+  const listeners = process.listeners(event);
+  for (const listener of listeners) {
+    process.off(event, listener);
+  }
+  return () => {
+    for (const listener of listeners) {
+      process.on(event, listener);
+    }
+  };
+};
+
+const makeUnhandledTracker = () => {
+  let unhandled = 0;
+  let seenUnhandled = 0;
+  const pending = new Set();
+  const tokenByPromise = new WeakMap();
+  // Finalization callbacks run on the main thread; `pending` is only mutated
+  // here and after the flushes in countUnhandled, so no extra synchronization
+  // is required.
+  const registry = new FinalizationRegistry(token => {
+    if (pending.delete(token)) {
+      // We count once per promise when it is collected, regardless of whether
+      // a handler was attached later.
+      unhandled += 1;
+    }
+  });
+
+  const onUnhandled = (_reason, promise) => {
+    seenUnhandled += 1;
+    const token = Symbol('unhandledRejection');
+    tokenByPromise.set(promise, token);
+    pending.add(token);
+    registry.register(promise, token, token);
+  };
+
+  const onHandledLate = promise => {
+    tokenByPromise.delete(promise);
+    // Keep the token in `pending` so late handlers don't mask an earlier
+    // unhandled notification; we want to count every unhandledRejection event
+    // that fired in the subprocess even if it is observed later. The registry
+    // callback runs at most once per promise on the main thread, so this does
+    // not double-count. Tokens stay in `pending` only until the promise is GC'd
+    // and the finalizer fires to delete them.
```

Comment:
This is not the right semantics for how HardenedJS behaves.  Please see https://github.com/endojs/endo/tree/master/packages/ses/src/error/unhandled-rejection.js for how the event handlers keep track of late handling without counting those as unhandled rejections (as opposed to what you have in `makeUnhandledTracker`).